proxyclient/client.go

```diff
@@ -0,0 +1,239 @@
+package proxyclient
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/gorilla/websocket"
+	"github.com/rancher/remotedialer"
+	"github.com/rancher/wrangler/v3/pkg/generated/controllers/core"
+	v1 "github.com/rancher/wrangler/v3/pkg/generated/controllers/core/v1"
+
+	"github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/cache"
+)
+
+const (
+	defaultServerAddr        = "wss://127.0.0.1"
+	defaultServerPort        = 5555
+	defaultServerPath        = "/connect"
+	retryTimeout             = 1 * time.Second
+	certificateWatchInterval = 10 * time.Second
+)
+
+type PortForwarder interface {
+	Start() error
+	Stop()
+}
+
+type ProxyClientOpt func(*ProxyClient)
+
+type ProxyClient struct {
+	forwarder           PortForwarder
+	serverUrl           string
+	serverConnectSecret string
+
+	dialer    *websocket.Dialer
+	dialerMtx sync.Mutex
+
+	secretController v1.SecretController
+	namespace        string
+	certSecretName   string
+	certServerName   string
+
+	onConnect func(ctx context.Context, session *remotedialer.Session) error
+}
+
+func New(ctx context.Context, serverSharedSecret, namespace, certSecretName, certServerName string, restConfig *rest.Config, forwarder PortForwarder, opts ...ProxyClientOpt) (*ProxyClient, error) {
+	if restConfig == nil {
+		return nil, fmt.Errorf("restConfig required")
+	}
+
+	if forwarder == nil {
+		return nil, fmt.Errorf("a PortForwarder must be provided")
+	}
+
+	if namespace == "" {
+		return nil, fmt.Errorf("namespace required")
+	}
+
+	if certSecretName == "" {
+		return nil, fmt.Errorf("certSecretName required")
+	}
+
+	if serverSharedSecret == "" {
+		return nil, fmt.Errorf("server shared secret must be provided")
+	}
+
+	serverUrl := fmt.Sprintf("%s:%d%s", defaultServerAddr, defaultServerPort, defaultServerPath)
+
+	client := &ProxyClient{
+		serverUrl:           serverUrl,
+		forwarder:           forwarder,
+		serverConnectSecret: serverSharedSecret,
+		certSecretName:      certSecretName,
+		certServerName:      certServerName,
+		namespace:           namespace,
+	}
+
+	if err := client.buildDialer(ctx, restConfig); err != nil {
+		return nil, fmt.Errorf("dialer build failed %w: ", err)
+	}
+
+	for _, opt := range opts {
+		opt(client)
+	}
+
+	return client, nil
+}
+
+func (c *ProxyClient) buildDialer(ctx context.Context, restConfig *rest.Config) error {
+	core, err := core.NewFactoryFromConfigWithOptions(restConfig, nil)
+	if err != nil {
+		return fmt.Errorf("build secret controller failed: %w", err)
+	}
```

Comment:
We typically try not to create too many controller factories in Rancher to avoid duplicating data (memory). Since this client is going to run in Rancher, and Rancher already has controllers / caches for secrets, let's reuse those. To do this, you can make it so that `New()` takes a SecretController, and when we integrate with Rancher, we'll pass the controller.